agent/php_execute.c

```diff
@@ -1126,6 +1111,17 @@ static inline void nr_php_execute_segment_end(
       || stacked->error) {
     nr_segment_t* s = nr_php_stacked_segment_move_to_heap(stacked TSRMLS_CC);
     nr_php_execute_segment_add_metric(s, metadata, create_metric);
+
+    /*
+     * Check if code level metrics are enabled in the ini.
+     * If they aren't, exit and don't add any attributes.
+     */
+    if (NRINI(code_level_metrics_enabled)) {
+      if (NULL == s->attributes) {
+        s->attributes = nr_attributes_create(s->txn->attribute_config);
+      }
+      nr_php_txn_add_code_level_metrics(s->attributes, metadata);
+    }
```

Comment:
What if:
1. `nr_php_txn_add_code_level_metrics` would be renamed to `nr_php_execute_segment_add_clm_attributes`. The latter name seems to match functions purpose - the agent is adding clm attributes from execute metadata to the segment. Also CLM attributes apply more to the segment/span rather than transaction as a whole, which the first name might suggest.
2. Have `nr_php_execute_segment_add_clm_attributes` take the `s` and `metadata` as arguments to be more in line with what it's actually doing - adding attributes to the segment rather than having addition of attributes split: `nr_attributes_create` is called outside but the attributes are updated inside. Let's couple what shouldn't be separated.
3. Define `nr_php_execute_segment_add_clm_attributes` only for PHP7+. This would simplify this block of code  [here](https://github.com/zsistla/newrelic-php-agent/blob/clm_to_dev/agent/php_txn.c#L1117-L1122).
5. Move the definition of `nr_php_execute_segment_add_clm_attributes` from php_txn.c to php_execute.c and make it static - no other code uses this function. This would make the code base more focused. Additionally this would make the extraction of `nr_php_execute_metadata_t` unnecessary.
After these changes the lines above could read like this:
```
  #if ZEND_MODULE_API_NO >= ZEND_7_0_X_API_NO /* PHP >= PHP7 */
    if (NRINI(code_level_metrics_enabled)) {
      nr_php_execute_segment_add_clm_attributes(s, metadata);
    }
#endif
```
You can have a look at a more complete suggestion [here](https://github.com/zsistla/newrelic-php-agent/pull/2).

Reply:
regarding point 1 and 2:
Originally, the parameters were attributes and metadata to allow it also to be called from here to provide CLM for files: https://github.com/newrelic/newrelic-php-agent/pull/506/files#diff-beedad7bbfc9687348e557b4752fbf6f54e4407a6e3177e8cabaaec7e3ffd625R1226

Additionally, the function needed access to a variable or function that it could only access while living in php_txn.

This is also why the attributes were created outside of the function because we didn't want to double create when adding to a txn instead of just a segment.

however, it's been refactored significantly including us no longer adding it to a filename/txn, so I'll see if it can be moved back and simplified.

Reply:
Thanks for explaining the reason why the code looks like it is. It's hard to see this background by only looking at this PR.

Reply:
Resolved by 3e7ec823504a22e68c094b947fd86af06ebd41d9. Thank you @zsistla!